### PR TITLE
nextcloud{28,29}Packages: update; revamp generation script

### PR DESCRIPTION
**Pull request by pyrox0**
## Description of changes
This updates the nextcloudPackages sets to use SRI hashes instead of sha256 hashes, and regenerates the package set using a new generation script that just uses bash, jq, and curl, as the previous generation script's tool, nc4nix, doesn't work properly anymore. This should make updating the package set in the future better and more robust. I've double-checked that the new hashes work properly, so this should not need to be reverted in future(as happened in #326912)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by dotlambda** [comment]
Is this better than using a more recent nc4nix and rewriting `fetchNextcloudApp` to use `fetchurl`?

**Comment by pyrox0** [comment]
I think this makes more sense since it doesn't require any non-core packages, and guarantees compatibility by using nix as the hasher. To me this made more sense than rewriting the fetcher, and I think rewriting the fetcher would be more trouble than it would be worth here, as you'd need to unpack it anyways. Leaving it as fetchzip means you don't have to handle that manually.

File: pkgs/servers/nextcloud/packages/generate.sh [hunk 1]
```diff
@@ -8,7 +8,55 @@ set -x
...
+    # Add a starting bracket to the apps file for this version
+    echo '{' >"$MAJOR_FILE".tmp
+    for a in ${APPS//,/ }; do
+        echo "Fetching $a"
+        # Ensure the app exists in the file
+        if [ "$(jq -r ".[] | select(.id == \"${a}\")" "$APPS_PER_VERSION")" != "" ]; then
+            # Get all of our variables
+            VERSION=$(jq -r ".[] | select(.id == \"${a}\") | .releases[0].version" "$APPS_PER_VERSION")
+            URL=$(jq -r ".[] | select(.id == \"${a}\") | .releases[0].download" "$APPS_PER_VERSION")
+            HASH=$(nix store prefetch-file --json --hash-type sha256 --unpack "$URL" | jq -r .hash)
```
Comment (Emantor) [inline pkgs/servers/nextcloud/packages/generate.sh:33]: This raises
```
[…]
++ jq -r .hash                                                                                                                                                                                    
++ nix store prefetch-file --json --hash-type sha256 --unpack https://github.com/nextcloud/bookmarks/releases/download/v14.2.2/bookmarks-14.2.2.tar.gz                                            
error: unrecognised flag '--unpack' 
```
when I try to run the generation script. How does this work? My nix version is `nix (Nix) 2.18.4
`.

**Comment by Ma27** [comment]
So, this change feels a bit off and I'm considering to propose a revert now.

Didn't the old update script ignore prereleases? I now get contacts 6.0 __rc1__ and a nightly release for maps.
Since this was never discussed here, I can assume that this is a regression?

That said, I don't think that a hacky bash script is the best option here. What is the current issue even?
Did someone even investigate what was wrong with nc4nix? There weren't any changes code-wise, so it might've been a force-pushed release?

Also, how do we want to proceed with stable backports? Given that Nextcloud discloses vulnerabilities a few weeks after the release, we should still backport Nextcloud. But now I have to do it manually because

* all hashes changed and there are conflicts
* we probably don't want nightly app releases on stable

cc @dotlambda @pyrox0

File: pkgs/servers/nextcloud/packages/generate.sh (hunk 1 above)
Comment (pyrox0) [reply pkgs/servers/nextcloud/packages/generate.sh:33]: Ah, --unpack is nix 2.20.0/lix-exclusive. I'll send a followup that uses nix-prefetch-url instead, which has a backwards-compatible --unpack option that nix 2.18 can use.

**Comment by pyrox0** [comment]
Looking at how nextcloud filters out pre-releases and nightly versions. Shouldn't be too hard, and will push a PR with updated packages and script.

**Comment by dotlambda** [comment]
> Didn't the old update script ignore prereleases? I now get contacts 6.0 **rc1** and a nightly release for maps. Since this was never discussed here, I can assume that this is a regression?

No, I've seen plenty of nightly versions with nc4nix in the past.

File: pkgs/servers/nextcloud/packages/generate.sh (hunk 1 above)
Comment (Emantor) [reply pkgs/servers/nextcloud/packages/generate.sh:33]: FWIW I just fixed this up locally by requiring `nixVersions.latest` in the `generate.sh` script.

Comment (Ma27) [reply pkgs/servers/nextcloud/packages/generate.sh:33]: I think we want to pin Nix to an explicit version here. Technically, you can still use Nix 2.3 to instantiate nixpkgs and requiring newer features is fine here, but it has to be made explicit.

**Comment by Ma27** [comment]
So, I did a bit of debugging:

`apps.json` from 29:

```
$ jq < apps.json '.[]|select(.id=="contacts").releases|.[]|.version'
"6.0.0-rc.1"
"6.0.0-alpha.1"
"6.0.0"
"5.5.3"
"5.5.2"
"5.5.1"
"5.5.0-rc.2"
"5.5.0"
```

We'll always get the RC here even though there's a 6.0.0 out already.

I'm pretty sure that's the case that https://github.com/helsinki-systems/nc4nix/pull/5 is fixing. I'm not sure though how nightly versions appeared here, perhaps a bug (cc @onny).

---

First and foremost, this is clearly a regression from this fetcher.
And I'll ask again: did anybody bother to investigate what was wrong with nc4nix? Have we ruled out that somebody did - against all best practices - force-push an app release?
I don't want to throw out a fetcher written in a real programming language over this.

**Comment by Pizmovc** [comment]
I see two possibilities for moving this forward:

1. We fix all issues with the current script (usage of non-stable releases, nix not being included in the nix-shell, ???)
2. We revert this MR and use a newer version of `nc4nix` which should've fixed the issue which preceded this PR

I would prefer the second option, since it places less maintenance burden on us. What are your thoughts on this?

File: pkgs/servers/nextcloud/packages/generate.sh [hunk 2]
```diff
@@ -8,7 +8,55 @@ set -x
...
+    curl "https://apps.nextcloud.com/api/v1/platform/${v}/apps.json" -o "$APPS_PER_VERSION"
+
+    # Add a starting bracket to the apps file for this version
+    echo '{' >"$MAJOR_FILE".tmp
+    for a in ${APPS//,/ }; do
+        echo "Fetching $a"
+        # Ensure the app exists in the file
+        if [ "$(jq -r ".[] | select(.id == \"${a}\")" "$APPS_PER_VERSION")" != "" ]; then
+            # Get all of our variables
+            VERSION=$(jq -r ".[] | select(.id == \"${a}\") | .releases[0].version" "$APPS_PER_VERSION")
```
Comment (dotlambda) [inline pkgs/servers/nextcloud/packages/generate.sh:31]: This just picks the first item in releases, which might not actually be the most recent version. See e.g. the calendar app.

**Review by dotlambda — COMMENTED** [review]
I now agree we should either revert this or improve it significantly.
The former is what I do in https://github.com/NixOS/nixpkgs/pull/331507.

**Pull request merged**